ape_infura/provider.py

```diff
@@ -42,15 +60,20 @@ def uri(self) -> str:
         if (ecosystem_name, network_name) in self.network_uris:
             return self.network_uris[(ecosystem_name, network_name)]
 
-        key = None
-        for env_var_name in _ENVIRONMENT_VARIABLE_NAMES:
-            env_var = os.environ.get(env_var_name)
-            if env_var:
-                key = env_var
-                break
+        key = self.get_random_api_key()
 
-        if not key:
-            raise MissingProjectKeyError()
+        prefix = f"{ecosystem_name}-" if ecosystem_name != "ethereum" else ""
+        network_uri = f"https://{prefix}{network_name}.infura.io/v3/{key}"
+        self.network_uris[(ecosystem_name, network_name)] = network_uri
+        return network_uri
+
+    def get_new_uri(self) -> str:
+        """
+        To generate a new URI with a new API key. Added to keep backwards compatibity
```

Comment:
question: backwards compat with what exactly?

Reply:
Actually, if we update the `URI` with the logic, the tests are failing, so I thought to make a separate method.